Source/RunActivity/Viewer3D/Shapes.cs

```diff
@@ -2605,45 +2669,82 @@ public void PrepareFrame(RenderFrame frame, WorldPosition location, Matrix[] ani
         {
-            int[] h = LodControls[0].DistanceLevels[0].SubObjects[0].ShapePrimitives[0].Hierarchy;
-            Matrix matrix = Matrix.Identity;
-            while (iNode != -1)
+            bones = null; // standard scenery material has no skin and bones
+            var xnaMatrix = Matrix.Identity;
+            var hi = shapePrimitive.HierarchyIndex;
+            while (hi >= 0 && hi < shapePrimitive.Hierarchy.Length)
             {
-                matrix *= Matrices[iNode];
-                iNode = h[iNode];
+                Matrix.Multiply(ref xnaMatrix, ref animatedXNAMatrices[hi], out xnaMatrix);
+                hi = shapePrimitive.Hierarchy[hi];
             }
-            return matrix;
+            Matrix.Multiply(ref xnaMatrix, ref xnaDTileTranslation, out xnaMatrix);
+            return xnaMatrix;
         }
 
-        public int GetParentMatrix(int iNode)
+        public virtual Matrix GetMatrixProduct(int iNode)
         {
-            return LodControls[0].DistanceLevels[0].SubObjects[0].ShapePrimitives[0].Hierarchy[iNode];
+            var h = GetModelHierarchy();
+            Matrix matrix = Matrix.Identity;
+            if (h != null && h.Length > iNode)
+            {
+                while (iNode != -1)
+                {
+                    matrix *= Matrices[iNode];
+                    iNode = h[iNode];
+                }
+            }
+            return matrix;
         }
 
+        public int[] GetModelHierarchy() => LodControls?.FirstOrDefault()?.DistanceLevels?.FirstOrDefault()?.SubObjects?.FirstOrDefault()?.ShapePrimitives?.FirstOrDefault()?.Hierarchy;
+
+        /// <summary>
+        /// This method is part of the animation handling. Gets the parent that will be animated, for finding a bogie for wheels.
+        /// </summary>
+        public int GetParentMatrix(int iNode) => GetModelHierarchy()?.ElementAtOrDefault(iNode) ?? -1;
+
+        /// <summary>
+        /// Searches for the parent animation.
+        /// </summary>
+        /// <param name="animationId">For stf files it is the node id, for gltf files it is the animation id.</param>
+        /// <returns>The parent animation id.</returns>
+        public virtual int GetAnimationParent(int animationId) => GetParentMatrix(animationId);
+
+        /// <summary>
+        /// Tells whether the animation is an internal sequence defined within the shape, or is just a tag that needs external animation.
+        /// </summary>
+        /// <param name="animationId">For stf files it is the node id, for gltf files it is the animation id.</param>
+        /// <returns>true if there is no internal seqence defined in the shape.</returns>
+        public virtual bool IsAnimationArticulation(int animationId) => Animations?.FirstOrDefault()?.anim_nodes?.ElementAtOrDefault(animationId)?.controllers.Count == 0;
```

Comment:
See my other comment for context on this.

```suggestion
        public virtual bool IsAnimationArticulation(int animationId) => Animations?.FirstOrDefault()?.anim_nodes?.ElementAtOrDefault(animationId)?.controllers.Count > 0;
```

Reply:
I don't think this change is correct, see above.

Reply:
Changing the == to > is for sure wrong. But I wonder if the original check's opposite and the new method is somehow not equivalent.

The opposite of the original:

`!(TrainCarShape.SharedShape.Animations != null && TrainCarShape.SharedShape.Animations.Count > 0 && TrainCarShape.SharedShape.Animations[0].anim_nodes.Count > matrix && TrainCarShape.SharedShape.Animations[0].anim_nodes[matrix].controllers.Count > 0)`

The new:

`Animations?.FirstOrDefault()?.anim_nodes?.ElementAtOrDefault(animationId)?.controllers.Count == 0`

Reply:
The only check that is missing from the new method is this: `Animations[0].anim_nodes.Count > matrix`. This might be the key?

Reply:
Oops, and also the null check might be missing. How about this:

`       public virtual bool IsAnimationArticulation(int animationId) =>
           Animations?.FirstOrDefault()?.anim_nodes is anim_nodes a && a.Count <= animationId && a.ElementAtOrDefault(animationId)?.controllers is controllers c && c.Count == 0;
`
Unfortunately I cannot check it now.

Reply:
> My man, you are overcomplicating this. You realize (...)

My man, you are oversimplifying this. :-) You realize, that the whole point moving this check into its own method is to be able to override it for the gltf shapes, right? (The animation handling is somewhat different.) Shuffling around like you suggested requires to negate that equation as well. But the whole thing is not needed, I just failed to correctly replicate the original equation in line 468, what you mentioned in your first post here, which is now fixed. Thanks for finding this bug, it was a really  good catch!

Reply:
Well from my perspective glTF doesn't exist yet, nothing I got uses it! If the glTF setup can't fit in the same place then it's fine for it to be split out into a method to be chosen depending on the type of SharedShape used.

Anyway, I'll check again but it does look like the new implementation will fix the null-animation case with the `is` checks and checking for > 0 controllers. Though now that you negate that and then negate it again...that's 2/3 places it's negated, might as well remove the negation from Shapes.cs and MSTSWagonViewer.cs, but then swap the glTF implementation from == null to != null. Little less confusing that way.

Reply:
> Well from my perspective glTF doesn't exist yet
Okay, but we are talking under literally the glTF PR. :-) 

> Though now that you negate that and then negate it again...that's 2/3 places it's negated, might as well remove the negation
Fine, I will do it for you, as a reward for the lot of text, but it will have to wait a couple of days. Btw this PR is HUGE, so if you look thoroughly, may find a lot more optimisation possibilities probably. You found this one only because there was a bug here. But these optimisations are no big deal, mostly just coding style...